<a href="https://colab.research.google.com/github/Lalit-V/aiag/blob/main/aiagents.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
!pip install -q crewai # -q => quietly install

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 kB 685.8 kB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 67.3/67.3 kB 5.0 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.3/133.3 kB 9.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.2/50.2 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 210.9/210.9 kB 14.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 997.8/997.8 kB 37.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 362.9/362.9 kB 21.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.5/61.5 kB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.5/52.5 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 109.5/109.5 kB 7.8 MB/

In [3]:
!pip install -q crewai[tools]

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 462.7/462.7 kB 12.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.8/147.8 kB 7.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.1/115.1 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.6/21.6 MB 43.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 341.8/341.8 kB 14.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.6/57.6 kB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.4/9.4 MB 52.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 476.0/476.0 kB 17.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.7/98.7 kB 4.3 MB/s eta 0:00:00


In [4]:
!pip install -q python-dotenv # to securely access api-keys. Never embed bare api-keys

In [7]:
%reload_ext dotenv
%dotenv

In [8]:
import os
from crewai import Agent, Task, Crew, Process
from crewai_tools import SerperDevTool
from google.colab import userdata

llm_key = userdata.get('llmApiKi')
google_search_key = userdata.get('srprApiKi')

/usr/local/lib/python3.10/dist-packages/pydantic/_internal/_config.py:341: UserWarning: Valid config keys have changed in V2:
* 'allow_population_by_field_name' has been renamed to 'populate_by_name'
* 'smart_union' has been removed
  warnings.warn(message, UserWarning)


In [9]:
search_tool = SerperDevTool()

In [10]:
from langchain_cohere import ChatCohere

# Let's Define AI Agents' Role and Goals

## *Researcher Agent*

In [15]:
researcher = Agent (
    role="Sr. Research Assistant",
    goal="Research latest advancements in AI agents' technology.",
    backstory="""You work at a leading AI resesarch organization. You are
    an expert in searching internet for AI Agent Frameworks. You excel in
    analyzing complex information and generating actionable insights from it.""",
    verbose=True,
    allow_delegation=False,
    tools=[search_tool],
    llm=ChatCohere(model='command-r-plus', temperature=0.2, cohere_api_key=llm_key)
)

## *Writer Agent*

In [16]:
writer = Agent(
    role="Information Strategist",
    goal="Write a summary of the latest advacements in AI Agent technology and list key highlights.",
    backstory="""You are a famous AI Agent content strategist, known for your
    knowledge insights and engaging articles. You transform complex concepts
    into compelling narratives.""",
    verbose=True,
    allow_delegation=True,
    llm=ChatCohere(model='command-r-plus', temperature=0.7, cohere_api_key=llm_key)
)

## **Give Agents Tasks to Perform**

In [14]:
netkokhangalo = Task(
    description="""Conduct a comprehensive analysis of the latest advancements
    in AI Agent technology and applications. Identify major trends, breakthroughs,
    and potential economic impact on life sciences industry.""",
    expected_output="Full analysis report in bullet points.",
    agent=researcher
)

topdogeditor = Task(
    description="""Using the researched insights provided to you, develop an
    executive report that hightlight the most significant AI Agents' advancements.
    Your audience is C-suite and board memebers of a large healthcare network. Your
    report should be highly professional and excellent quality. Avoid complex
    words and make sure it does not sound like that it is AI generated""",
    expected_output="One page report with summary paragraph, key highlights, and take home message.",
    agent=writer
)

## **Time to put the Crew to Work!**

In [17]:
crew = Crew(
    agents=[researcher, writer],
    tasks=[topdogeditor, topdogeditor],
    verbose=1, # 1 0r 2 for different logging levels.
)
result = crew.kickoff()
print("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
print(result)
print("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")

 [2024-08-21 16:29:04][DEBUG]: == Working Agent: Sr. Research Assistant
 [2024-08-21 16:29:04][INFO]: == Starting Task: Conduct a comprehensive analysis of the latest advancements 
    in AI Agent technology and applications. Identify major trends, breakthroughs, 
    and potential economic impact on life sciences industry.
 [2024-08-21 16:30:05][DEBUG]: == [Sr. Research Assistant] Task output: Here is a comprehensive analysis of the latest advancements in AI Agent technology, along with identified major trends, breakthroughs, and potential economic impacts on the life sciences industry:

- **Latest AI Agent Advancements**:
   - AI agents continue to evolve with advancements in underlying technologies such as deep learning, reinforcement learning, and neural language models.
   - GPT-4o, Project Astra, and Auto-GPT are some of the leading AI agents that represent a significant leap forward in terms of capabilities.
   - There is a growing trend towards specialization and multi-agent de

# Executive Report: AI Agent Technology Advancements

## Summary
The latest advancements in AI Agent technology are transformative, with a focus on generative capabilities and specialized applications. AI agents are now powered by deep learning, reinforcement learning, and neural language models, enabling human-like interactions and content generation. This evolution is set to revolutionize the healthcare industry, with improved diagnostics and therapeutics, increased productivity, and accelerated innovation. The economic impact of these advancements is significant, with the AI in Life Sciences Market projected to reach USD 8.88 billion by 2029 and generative AI expected to contribute up to $110 billion annually.

## Key Highlights
- **Technological Advancements:** Deep learning and neural language models drive the next wave of AI agents, with Large Language Models (LLMs) at the core.
- **Generative AI:** AI systems can now generate content, reshaping creative processes and offering new opportunities for enterprises.
- **Transformational Impact:** AI agents like GPT-4o and Project Astra demonstrate real-time conversational capabilities, blurring the line between humans and machines.
- **Specialization and Collaboration:** There is a growing trend towards tailoring AI agents for specific tasks and enabling them to work collaboratively.
- **Economic Value:** The life sciences industry is witnessing a growing adoption of AI, with generative AI expected to have a potential economic impact of $60-$110 billion annually.

## Take-Home Message
The latest advancements in AI Agent technology are set to revolutionize the healthcare industry and beyond. AI agents are now faster, more intelligent, and capable of human-like interactions, with a focus on generative capabilities. The economic impact of these advancements is significant, creating new opportunities for innovation and improved patient care. The future of AI in healthcare is promising, and these breakthroughs will shape the way we interact with technology and each other.

---

This executive report provides a concise yet comprehensive overview of the latest AI Agent technology advancements and their potential impact on the healthcare industry. It highlights the transformative nature of these advancements and the economic value they bring, offering a glimpse of the exciting future of AI in healthcare.
